libass/ass_render.c

```diff
@@ -990,25 +1076,122 @@ static void fill_composite_hash(CompositeHashKey *hk, CombinedBitmapInfo *info)
 static void
 get_outline_glyph(ASS_Renderer *priv, GlyphInfo *info)
 {
-    memset(&info->hash_key, 0, sizeof(info->hash_key));
+    OutlineHashValue *val;
+    ASS_DVector scale, offset = {0};
 
+    int32_t asc, desc;
     OutlineHashKey key;
-    fill_glyph_hash(priv, &key, info);
-    OutlineHashValue *val = ass_cache_get(priv->cache.outline_cache, &key, priv);
-    if (!val || !val->valid)
-        return;
+    if (info->drawing_text) {
+        key.type = OUTLINE_DRAWING;
+        key.u.drawing.text = info->drawing_text;
+        val = ass_cache_get(priv->cache.outline_cache, &key, priv);
+        if (!val || !val->valid) {
+            ass_cache_dec_ref(val);
+            return;
+        }
+
+        double w = priv->font_scale / (1 << (info->drawing_scale - 1));
+        scale.x = info->scale_x * w;
+        scale.y = info->scale_y * w;
+        desc = 64 * info->drawing_pbo;
+        asc = val->asc - desc;
+
+        offset.y = -asc * scale.y;
+    } else {
+        key.type = OUTLINE_GLYPH;
+        GlyphHashKey *k = &key.u.glyph;
+        k->font = info->font;
+        k->size = info->font_size;
+        k->face_index = info->face_index;
+        k->glyph_index = info->glyph_index;
+        k->bold = info->bold;
+        k->italic = info->italic;
+        k->flags = info->flags;
+        if (info->font->desc.vertical && info->symbol >= VERTICAL_LOWER_BOUND)
+            k->flags |= DECO_ROTATE;
+
+        val = ass_cache_get(priv->cache.outline_cache, &key, priv);
+        if (!val || !val->valid) {
+            ass_cache_dec_ref(val);
+            return;
+        }
+
+        scale.x = info->scale_x;
+        scale.y = info->scale_y;
+        asc  = val->asc;
+        desc = val->desc;
+    }
+
+    info->outline = val;
+    info->outline_transform.scale = scale;
+    info->outline_transform.offset = offset;
+
+    info->bbox.x_min = lrint(val->cbox.x_min * scale.x + offset.x);
+    info->bbox.y_min = lrint(val->cbox.y_min * scale.y + offset.y);
+    info->bbox.x_max = lrint(val->cbox.x_max * scale.x + offset.x);
+    info->bbox.y_max = lrint(val->cbox.y_max * scale.y + offset.y);
 
-    info->hash_key.u.outline.outline = val;
-    info->outline = &val->outline;
-    info->border[0] = &val->border[0];
-    info->border[1] = &val->border[1];
-    info->bbox = val->bbox_scaled;
     if (info->drawing_text || priv->settings.shaper == ASS_SHAPING_SIMPLE) {
-        info->cluster_advance.x = info->advance.x = val->advance.x;
-        info->cluster_advance.y = info->advance.y = val->advance.y;
+        info->cluster_advance.x = info->advance.x = lrint(val->advance * scale.x);
+        info->cluster_advance.y = info->advance.y = 0;
```

Comment:
Ah, I see that you remove `advance.y` entirely in the main commit. IIRC this is because
1. drawings always have `advance.y = 0`, which by this point is clear from our code because you explicitly initialize it to zero, and
2. FreeType glyphs also always have `advance.y = 0` because FreeType promises this in horizontal writing and that’s exactly what we request from it; however, _this,_ in my opinion, is _not_ obvious from our current code or your commit. Splitting this change off into a separate commit would be ideal but doesn’t seem easy at all, but even when it is contained in this big commit, you should explain this situation in the commit message.